### PR TITLE
Add OpenSSL

### PR DESCRIPTION
**Pull request by kroeckx**
I was unable to test this because of various issues.

### CODE REVIEW

**Comment by mikea** [comment]
Thanks for adding this. I was actually in the process of trying to fix OpenSSL fuzzers:

https://github.com/openssl/openssl/pull/1788

Is this the same problem you had?

**Comment by kroeckx** [comment]
No, the issues I have is that:
- I don't get docker working locally, something seems to be wrong with the package in Debian.
- When I build and run it using libfuzzer (which I haven't tried before), I get:
````
ERROR: __sanitizer_set_death_callback is not defined. Exiting.
Did you use -fsanitize-coverage=... to build your code?
````

**Comment by kcc** [comment]
did you also use -fsanitize=address or other sanitizer? 
( I need to change the error message to 
`Did you use -fsanitize-coverage=...  and -fsanitize=[address,undefined,...] to build your code?`)

**Comment by kroeckx** [comment]
Oh, I did a build without -fsanitize=address. But I understand you do put that in the CFLAGS, so it should work.

File: targets/openssl/build.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+cd /src/openssl
+
+./config enable-fuzz-libfuzzer -DPEDANTIC no-shared --with-fuzzer-lib=/usr/lib/libfuzzer $CFLAGS
+LDCMD=clang++ make
```
Comment (mikea) [inline targets/openssl/build.sh]: -j$(nproc) please

Comment (mikea) [inline targets/openssl/build.sh]: for some reason it doesn't build with LDCMD for me, but builds totally fine without setting it at all. Why did you need it?

File: targets/openssl/build.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+cd /src/openssl
+
+./config enable-fuzz-libfuzzer -DPEDANTIC no-shared --with-fuzzer-lib=/usr/lib/libfuzzer $CFLAGS
```
Comment (mikea) [inline targets/openssl/build.sh]: Right now it is /work/lib/libfuzzer
btw, do you think we should move it to /usr/lib instead?

Comment (mikea) [inline targets/openssl/build.sh]: unfortunately just $CFLAGS doesn't work due to the pain you have to endure to produce static binaries with clang :) 

We need to pass $FUZZER_LDFLAGS to linker. Maybe through LDCMD?

Comment (kroeckx) [inline targets/openssl/build.sh]: The documentation says it's in /usr/lib ...

File: targets/openssl/build.sh (hunk 1 above)
Comment (kroeckx) [inline targets/openssl/build.sh]: Can I suggest you add that to the template?

File: targets/openssl/build.sh (hunk 2 above)
Comment (kroeckx) [inline targets/openssl/build.sh]: From the logs of others, I see it's setting that to: ```-Wl,-whole-archive /usr/local/lib/libc++.a /usr/local/lib/libc++abi.a -Wl,-no-whole-archive -L/work/lib```

But that makes no sense to me. It's like you're linking a C++ program with a C compiler instead of the C++ compiler. The compiler really should add that itself.

File: targets/openssl/build.sh (hunk 1 above)
Comment (kroeckx) [inline targets/openssl/build.sh]: So that it links and doesn't throw a bunch of errors about functions in libfuzzer.a that it can't find, see comment above.

File: targets/openssl/build.sh (hunk 2 above)
Comment (kroeckx) [inline targets/openssl/build.sh]: You want static binaries? Is using the shared libraries like pthread, dl, m, c, stdc++ a problem?

Comment (kroeckx) [inline targets/openssl/build.sh]: So is the binary /work/lib/libfuzzer.a or /work/lib/libfuzzer/ibfuzzer.a? I'm assuming the first.

Comment (mikea) [inline targets/openssl/build.sh]: We want "mostly" static binaries (fully static would work too, but it is even trickier to produce them). pthread, dl, m, c are ok. The rest (including stdc++) is a problem.

I agree, this ideally should be fixed in the clang driver. To be fair we have built llvm with `-DLIBCXX_ENABLE_SHARED=OFF -DLIBCXX_ENABLE_STATIC_ABI_LIBRARY=ON` which might not be fully supported in the driver.

File: targets/openssl/build.sh (hunk 1 above)
Comment (mikea) [inline targets/openssl/build.sh]: done.

File: targets/openssl/build.sh (hunk 2 above)
Comment (kroeckx) [inline targets/openssl/build.sh]: So the binaries I produced locally contained:
```
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

Without linking them to libfuzzer I get this instead:
```
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

And just using clang++ instead of clang gets me:
```
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

File: targets/openssl/build.sh (hunk 1 above)
Comment (mikea) [inline targets/openssl/build.sh]: Yeah, but then it fails with: 

```
LD_LIBRARY_PATH=.: clang++ -DDSO_DLFCN -DHAVE_DLFCN_H -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DPEDANTIC -DOPENSSLDIR="/usr/local/ssl" -DENGINESDIR="/usr/local/lib/engines-1.1" -Wall -O3 -pthread -m64 -DL_ENDIAN -Wextra -Qunused-arguments -fsanitize-coverage=edge,indirect-calls -fno-omit-frame-pointer -g -g -fsanitize=address -fsanitize-coverage=edge,indirect-calls,8bit-counters -o apps/openssl apps/app_rand.o apps/apps.o apps/asn1pars.o apps/ca.o apps/ciphers.o apps/cms.o apps/crl.o apps/crl2p7.o apps/dgst.o apps/dhparam.o apps/dsa.o apps/dsaparam.o apps/ec.o apps/ecparam.o apps/enc.o apps/engine.o apps/errstr.o apps/gendsa.o apps/genpkey.o apps/genrsa.o apps/nseq.o apps/ocsp.o apps/openssl.o apps/opt.o apps/passwd.o apps/pkcs12.o apps/pkcs7.o apps/pkcs8.o apps/pkey.o apps/pkeyparam.o apps/pkeyutl.o apps/prime.o apps/rand.o apps/rehash.o apps/req.o apps/rsa.o apps/rsautl.o apps/s_cb.o apps/s_client.o apps/s_server.o apps/s_socket.o apps/s_time.o apps/sess_id.o apps/smime.o apps/speed.o apps/spkac.o apps/srp.o apps/ts.o apps/verify.o apps/version.o apps/x509.o -L. -lssl -L. -lcrypto -ldl
/usr/bin/ld: cannot find -lstdc++
```

**Comment by kroeckx** [comment]
I pushed a new commit

File: targets/openssl/build.sh (hunk 2 above)
Comment (mikea) [inline targets/openssl/build.sh]: Because libstdc++.so.6 comes from clang (which is built from head), it has to be statically linked.

Comment (kroeckx) [inline targets/openssl/build.sh]: I think my commit should address this problem.

**Comment by mikea** [comment]
Works like a charm. Thank you. Submitting.

**Comment by mikea** [comment]
This was successfully built on a bot and it is on its way to ClusterFuzz.

**Comment by kroeckx** [comment]
Thanks!

**Comment by kroeckx** [comment]
So I have a question about this: does it each time start from the seed file again, or does it keep a corpus internally somewhere? If I update the corpus (in out git branch), will it pick it up?

**Comment by mikea** [comment]
It does keep a corpus internally. And yes, it should pick up new seeds automatically once you update them in the git.

**Comment by kroeckx** [comment]
If I look at the log file, it seems like it doesn't have any corpus at all, and that it's still using an old version of the repository?

**Comment by oliverchang** [comment]
Which log file are you looking at?

**Comment by kroeckx** [comment]
https://oss-fuzz-build-logs.storage.googleapis.com/build_logs/openssl/latest.txt

**Comment by kroeckx** [comment]
I guess I might also just misunderstand what is going on.

**Comment by oliverchang** [comment]
Ah, those are our builder logs, which do some basic testing (running the fuzzers) before uploading them, so no corpus is passed. We do have logs for our actual fuzzing, but those aren't public just yet.

What indicated to you that we're using an old version of the repos? Based on the git checkout steps in https://oss-fuzz-build-logs.storage.googleapis.com/build_logs/openssl/latest.txt, both  oss-fuzz and openssl repos appear to be up to date.

oss-fuzz: 122733498ac7655804a0bdb3b2dc002846f1f533
openssl: 8aefa08cfbc7db7cc10765ee9684090e37983f45

**Comment by kroeckx** [comment]
It has this:

    fuzz/conf.c:27: OpenSSL internal error: assertion failed: (size_t)BIO_write(in, buf, len) == len

It happens with 0 sized files.

@mikea just fixed this for the cms and server fuzzer and said it where the only 2, but obviously conf has the same issue.

**Comment by mikea** [comment]
Looks like I introduced a bug to build functionality (it should fail the build) and missed this case. I'll fix the build and will send a PR to openssl.

**Comment by mikea** [comment]
https://github.com/openssl/openssl/pull/1829
